File: BridgeWeb/src/core/models/bridge-resource-descriptor.ts
import { z } from 'zod';

import {
	bridgeProtocolIdSchema,
	bridgeResourceKindSchema,
	type BridgeProtocolId,
	type BridgeResourceKind,
} from './bridge-core-models.js';

export const bridgeIntegrityDescriptorSchema = z.discriminatedUnion('kind', [
	z
		.object({
			kind: z.literal('wholeHash'),
			algorithm: z.enum(['sha256']),
			value: z.string().min(1),
		})
		.strict(),
	z
		.object({
			kind: z.literal('chunkManifest'),
			algorithm: z.enum(['sha256']),
			manifestResourceId: z.string().min(1),
		})
		.strict(),
	z
		.object({
			kind: z.literal('previewOnly'),
		})
		.strict(),
]);

export const bridgeIdentitySchema = z
	.object({
		paneId: z.string().min(1),
		protocol: bridgeProtocolIdSchema,
		sourceId: z.string().min(1).optional(),
		packageId: z.string().min(1).optional(),
		generation: z.number().int().nonnegative().optional(),
		revision: z.number().int().nonnegative().optional(),
		streamId: z.string().min(1).optional(),
		cursor: z.string().min(1).optional(),
	})
	.strict();

export const bridgeResourceDescriptorSchema = z
	.object({
		descriptorId: z.string().min(1),
		protocol: bridgeProtocolIdSchema,
		resourceKind: bridgeResourceKindSchema,
		resourceUrl: z.string().min(1),
		identity: bridgeIdentitySchema,
		content: z
			.object({
				mediaType: z.string().min(1),
				encoding: z.enum(['utf-8', 'binary']).optional(),
				expectedBytes: z.number().int().nonnegative().optional(),
				maxBytes: z.number().int().positive(),
				integrity: bridgeIntegrityDescriptorSchema.optional(),
			})
			.strict(),
		window: z
			.object({
				start: z.number().int().nonnegative().optional(),
				count: z.number().int().positive().optional(),
				maxCount: z.number().int().positive(),
			})
			.strict()
			.optional(),
	})
	.strict();

export const bridgeDescriptorRefSchema = z
	.object({
		descriptorId: z.string().min(1),
		expectedProtocol: bridgeProtocolIdSchema,
		expectedResourceKind: bridgeResourceKindSchema,
		expectedIdentity: bridgeIdentitySchema,
	})
	.strict();

export const bridgeAttachedResourceDescriptorSchema = z
	.object({
		ref: bridgeDescriptorRefSchema,
		descriptor: bridgeResourceDescriptorSchema,
	})
	.strict();

export type { BridgeProtocolId, BridgeResourceKind };
export type BridgeIntegrityDescriptor = z.infer<typeof bridgeIntegrityDescriptorSchema>;
export type BridgeIdentity = z.infer<typeof bridgeIdentitySchema>;
export type BridgeResourceDescriptor = z.infer<typeof bridgeResourceDescriptorSchema>;
export type BridgeDescriptorRef = z.infer<typeof bridgeDescriptorRefSchema>;
export type BridgeAttachedResourceDescriptor = z.infer<
	typeof bridgeAttachedResourceDescriptorSchema
>;

File: Sources/AgentStudio/Features/Bridge/Models/Transport/BridgeResourceDescriptor.swift
import Foundation

struct BridgeIntegrityDescriptor: Codable, Equatable, Sendable {
    enum Kind: String, Codable, Equatable, Sendable {
        case wholeHash
        case chunkManifest
        case previewOnly
    }

    let kind: Kind
    let algorithm: String?
    let value: String?
    let manifestResourceId: String?
}

struct BridgeResourceIdentity: Codable, Equatable, Sendable {
    let paneId: String
    let protocolId: String
    let sourceId: String?
    let packageId: String?
    let generation: Int?
    let revision: Int?
    let streamId: String?
    let cursor: String?

    enum CodingKeys: String, CodingKey {
        case paneId
        case protocolId = "protocol"
        case sourceId
        case packageId
        case generation
        case revision
        case streamId
        case cursor
    }
}

struct BridgeResourceContentDescriptor: Codable, Equatable, Sendable {
    enum Encoding: String, Codable, Equatable, Sendable {
        case utf8 = "utf-8"
        case binary
    }

    let mediaType: String
    let encoding: Encoding?
    let expectedBytes: Int?
    let maxBytes: Int
    let integrity: BridgeIntegrityDescriptor?
}

struct BridgeResourceWindowDescriptor: Codable, Equatable, Sendable {
    let start: Int?
    let count: Int?
    let maxCount: Int
}

struct BridgeResourceDescriptor: Codable, Equatable, Sendable {
    let descriptorId: String
    let protocolId: String
    let resourceKind: String
    let resourceUrl: String
    let identity: BridgeResourceIdentity
    let content: BridgeResourceContentDescriptor
    let window: BridgeResourceWindowDescriptor?

    enum CodingKeys: String, CodingKey {
        case descriptorId
        case protocolId = "protocol"
        case resourceKind
        case resourceUrl
        case identity
        case content
        case window
    }
}

struct BridgeDescriptorRef: Codable, Equatable, Sendable {
    let descriptorId: String
    let expectedProtocol: String
    let expectedResourceKind: String
    let expectedIdentity: BridgeResourceIdentity
}

struct BridgeAttachedResourceDescriptor: Codable, Equatable, Sendable {
    let ref: BridgeDescriptorRef
    let descriptor: BridgeResourceDescriptor
}
